QuickFIXn/Fields/Converters/DateTimeConverter.cs

```diff
@@ -9,197 +11,620 @@ namespace QuickFix.Fields.Converters;
+    }
+
+    [DoesNotReturn]
+    private static DateTime ThrowDateTime(ReadOnlySpan<char> s)
+    {
+        throw new FieldConvertError($"Could not convert \"{s}\" to DateTime");
+    }
+
+    [DoesNotReturn]
+    private static TimeOnly ThrowTimeOnly(ReadOnlySpan<char> s)
+    {
+        throw new FieldConvertError($"Could not convert \"{s}\" to TimeOnly");
     }
 
     /// <summary>
-    /// Check if string is TimeOnly and, if yes, convert to DateTime.
+    /// Converts the specified span to a <see cref="DateOnly"/>.
+    /// The span must be in the format "yyyyMMdd".
     /// </summary>
-    /// <param name="str"></param>
-    /// <returns></returns>
-    /// <exception cref="FieldConvertError"/>
-    public static DateTime ParseToTimeOnly(string str)
+    /// <param name="str">A span containing the characters that represent a date to convert.</param>
+    /// <returns>
+    /// A value representing the date in <paramref name="str"/>.
+    /// </returns>
+    /// <exception cref="FieldConvertError">The conversion cannot be performed successfully.</exception>
+    public static DateOnly ParseToDateOnly(ReadOnlySpan<char> str)
     {
-        try {
-            DateTime d = (str.Length > TIME_ONLY_MAXLENGTH_WITHOUT_NANOSECONDS)
-                ? TimeOnlyFromNanoString(str)
-                : DateTime.ParseExact(str, TIME_ONLY_FORMATS, DATE_TIME_CULTURE_INFO, DATE_TIME_STYLES);
-
-            return new DateTime(1980, 1, 1) + d.TimeOfDay;
-        }
-        catch (Exception e)
+        if (!DateOnly.TryParseExact(str, "yyyyMMdd", CultureInfo.InvariantCulture, DateTimeStyles.None, out DateOnly dateOnly))
         {
-            throw new FieldConvertError("Could not convert string (" + str + ") to TimeOnly: " + e.Message, e);
+            throw new FieldConvertError($"Could not convert string ({str} to DateOnly)");
         }
+
+        return dateOnly;
     }
 
     /// <summary>
-    /// Gets the nanoseconds component of the date represented by this instance truncated to 100 nanosecond resolution
+    /// Converts the specified string to a <see cref="DateTime"/>.
+    /// The string must be in the format "yyyyMMdd".
+    /// </summary>
+    /// <param name="str">A <see cref="string"/> containing the characters that represent a date to convert.</param>
+    /// <returns>
+    /// A value representing the date in <paramref name="str"/> with its time component equal to 00:00:00 and
+    /// its <see cref="DateTime.Kind"/> equal to <see cref="DateTimeKind.Unspecified"/>.
+    /// </returns>
+    /// <remarks>
+    /// This method calls <see cref="ParseToDateOnly(ReadOnlySpan{char})"/> which returns a <see cref="DateOnly"/>.
+    /// Consider calling the latter for correctness.
+    /// </remarks>
+    /// <exception cref="FieldConvertError">The conversion cannot be performed successfully.</exception>
+    public static DateTime ParseToDateOnly(string str) => ParseToDateOnly((ReadOnlySpan<char>)str).ToDateTime(default);
```

Comment:
Oh, _come on_!

I just went though thinking this possible weird overload situation with ParseToTimeOnly, only to realize it wasn't there... and then 10 minutes after resolving that I run into an *actual* weird overload situation right here!

Pfff... I'm going to make this better somehow.  Not sure yet how.

Reply:
Yeah that is awkward, I guess the new DateOnly-returning variant should have a different name - ParseToDate, ParseDateOnly ?

Could be confusing either way. One tool might be to apply `[EditorBrowsable(EditorBrowsableState.Never)]` on the old method - it means that it is still valid to be used but doesn't show up in auto-complete in the IDE. This doesn't affect the compiler binding though so would only be useful with a differently-named method.